src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -358,23 +370,139 @@ def setClean(self, wflow):
+        returned by WMStatsServer regarding the time of the last request status transition
+        :param wflow:      MSRuleCleaner workflow representation
+        :return:           Bool
+        """
+        archDelayExpired = False
+        currentTime = int(time.time())
+        threshold = self.msConfig['archiveDelayHours'] * 3600
+        outputMask = ['RequestTransition']
+        inputConditon = {'RequestName': wflow['RequestName']}
+        try:
+            wmsReqInfo = self.wmstatsSvc.getFilteredActiveData(inputConditon, outputMask)
+            # self.logger.debug("wmsReqInfo: \n%s", pformat(wmsReqInfo))
+            if wmsReqInfo and (currentTime - wmsReqInfo[0]["RequestTransition"][-1]["UpdateTime"]) > threshold:
+                archDelayExpired = True
+        except Exception as ex:
+            msg = "General Exception while checking the time of the last request status transition for wflow: %s : %s"
+            self.logger.exception(msg, wflow['RequestName'], str(ex))
+        return archDelayExpired
+
+    def setArchDelayExpired(self, wflow):
+        """
+        A function to set the 'IsArchDelayExpired' flag
+        """
+        wflow['IsArchDelayExpired'] = self._checkArchDelayExpired(wflow)
+        return wflow
+
+    def _archive(self, wflow):
+        """
+        A function to make the actual request status transition during the
+        archival step.
+        :param wflow:       A MSRuleCleaner workflow representation
+        :return:            Bool
+        """
+        archiveStatus = False
+        try:
+            self.reqmgr2.updateRequestStatus(wflow['RequestName'], wflow['TargetStatus'])
+            archiveStatus = True
+        except Exception as ex:
+            msg = "General Exception while trying status transition to: %s for wflow: %s : %s"
+            self.logger.exception(msg, wflow['TargetStatus'], ['RequestName'], str(ex))
+        return archiveStatus
+
     def archive(self, wflow):
         """
         Move the workflow to the proper archived status after checking
         the full cleanup status
         :param  wflow:      A MSRuleCleaner workflow representation
-        :param  archStatus: Target status to transition after archival
         :return wflow:
         """
-        # NOTE: check allowed status transitions with:
-        #       https://github.com/dmwm/WMCore/blob/5961d2229b1e548e58259c06af154f33bce36c68/src/python/WMCore/ReqMgr/DataStructs/RequestStatus.py#L171
+        # Make all the needed checks before trying to archive
         if not (wflow['IsClean'] or wflow['ForceArchive']):
             msg = "Not properly cleaned workflow: %s" % wflow['RequestName']
             raise MSRuleCleanerArchival(msg)
+        elif not wflow['TargetStatus']:
+            msg = "Could not determine which archival status to target for: %s" % wflow['RequestName']
+            raise MSRuleCleanerArchival(msg)
+        elif not wflow['IsLogDBClean']:
```

Comment:
I don't understand why raising exceptions here. Same for the two block belows (IsArchDelayExpired and enableRealMode). Shouldn't we simply log (if not already) an error and move to the next workflow?

Reply:
As I started to explain in the previous section we must exit the function in order to avoid actual status transition to one  of `*-archival` statuses. And the best way to do this is with an exception, because that way we can signal the upper level function not to update any of the relevant counters regarding the failing worklow.